File: Javascript(一个前端工程师的挣扎）/Backtracking/79.word_search.js
/**
 * Question 79. Word Search
 * Solved Strategy:
 * Date: 10/22/2018
 * Description:  The important takeaway is how to create 2D array in Javascript.
 * the way using fill in old way will make it reference to the first column
 * The proper way is
 * Array.from(Array(row_length),_=>Array(col_length).fill(inital_value));
 *
 */


/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var visited = [[]];

var exist = function(board, word) {
    if (board === null) return null;



    visited = Array.from(Array(board.length), _ => Array(board[0].length).fill(false));

    //console.log(visited);

    for (let i = 0; i < board.length; i++) {
        for (let j = 0; j < board[0].length; j++) {
            if (board[i][j] === word.charAt(0)) {
                var res = dfs(board, word, i, j, 0);

                if (res) return res;
            }
        }
    }

    return false;
};

function dfs(board, word, i, j, cur) {

    if (cur === word.length) {
        //console.log("i was called");

        return true;
    }

    //console.log(visited);
    if (i < 0 || i >= board.length || j < 0 || j >= board[0].length) {
        return false;
    }
    if  (visited[i][j] || word.charAt(cur) != board[i][j]) {
//         console.log("I am in positon row: " + i + " col: " + j);
//         console.log("Current Checking: " + cur);
//         console.log("The cell I wanna check is " + visited[i][j]);

        return false;
    }
    else {
        visited[ i ][ j ] = true;  // marked as visited
        var flag = dfs(board, word, i + 1, j, cur + 1) ||
            dfs(board, word, i - 1, j , cur + 1) ||
            dfs(board, word, i, j + 1, cur + 1) ||
            dfs(board, word, i , j - 1, cur + 1);

        if (flag) return flag;

        visited[i][j] = false;

        return flag;
    }
}

File: Java/79.word_search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        if (word == null || word.length() == 0) return true;
        if (board == null || board.length == 0) return false;

        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (word.charAt(0) == board[i][j]) {
                    boolean[][] visited = new boolean[board.length][board[0].length];
                    if (dfsHelper(board, visited, word, i, j, 0)) return true;
                }
            }
        }
        return false;
    }

    public boolean dfsHelper(char[][] board, boolean[][] visited, String word, int row, int col, int curIdx) {
        if (curIdx == word.length()) return true;
        if (!validatePos(board, row, col)) return false;
        if (visited[row][col]) return false;
        if (word.charAt(curIdx) != board[row][col]) return false;

        visited[row][col] = true;
        boolean res = dfsHelper(board, visited, word, row - 1, col, curIdx + 1)
                || dfsHelper(board, visited, word, row + 1, col, curIdx + 1)
                || dfsHelper(board, visited, word, row, col + 1, curIdx + 1)
                || dfsHelper(board, visited, word, row, col - 1, curIdx + 1);
        visited[row][col] = false;
        return res;

    }

    public boolean validatePos(char[][] board, int r, int c) {
        if (r < 0 || r >= board.length || c < 0 || c >= board[0].length) return false;
        return true;
    }
}